# Step 0 Importing Libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

# Step 1 Importing Data and Data Preprocessing 

In [2]:
test_feat = pd.read_csv("test_set_features.csv")
train_label= pd.read_csv("training_set_labels.csv")
train_feat=pd.read_csv("training_set_features.csv")

In [3]:
train_label

respondent_id  xyz_vaccine  seasonal_vaccine
0                  0            0                 0
1                  1            0                 1
2                  2            0                 0
3                  3            0                 1
4                  4            0                 0
...              ...          ...               ...
26702          26702            0                 0
26703          26703            0                 0
26704          26704            0                 1
26705          26705            0                 0
26706          26706            0                 0

[26707 rows x 3 columns]

In [4]:
train_feat

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0                  0          1.0            0.0                        0.0   
1                  1          3.0            2.0                        0.0   
2                  2          1.0            1.0                        0.0   
3                  3          1.0            1.0                        0.0   
4                  4          2.0            1.0                        0.0   
...              ...          ...            ...                        ...   
26702          26702          2.0            0.0                        0.0   
26703          26703          1.0            2.0                        0.0   
26704          26704          2.0            2.0                        0.0   
26705          26705          1.0            1.0                        0.0   
26706          26706          0.0            0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                       0.0                   0.0                    0.0   
1                       1.0                   0.0                    1.0   
2                       1.0                   0.0                    0.0   
3                       1.0                   0.0                    1.0   
4                       1.0                   0.0                    1.0   
...                     ...                   ...                    ...   
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
0                              0.0                      1.0   
1                              0.0                      1.0   
2                              0.0                      0.0   
3                              1.0                      0.0   
4                              1.0                      0.0   
...                            ...                      ...   
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
0                        1.0  ...              Below Poverty     Not Married   
1                        1.0  ...              Below Poverty     Not Married   
2                        0.0  ...  <= $75,000, Above Poverty     Not Married   
3                        0.0  ...              Below Poverty     Not Married   
4                        1.0  ...  <= $75,000, Above Poverty         Married   
...                      ...  ...                        ...             ...   
26702                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26703                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26704                    1.0  ...                        NaN     Not Married   
26705                    NaN  ...  <= $75,000, Above Poverty         Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   

       rent_or_own   employment_status  hhs_geo_region  \
0              Own  Not in Labor Force        oxchjgsf   
1             Rent            Employed        bhuqouqj   
2              Own            Employed        qufhixun   
3             Rent  Not in Labor Force        lrircsnp   
4              Own            Employed        qufhixun   
...            ...                 ...             ...   
26702          Own  N

In [5]:
test_feat

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              26707          2.0            2.0                        0.0   
1              26708          1.0            1.0                        0.0   
2              26709          2.0            2.0                        0.0   
3              26710          1.0            1.0                        0.0   
4              26711          3.0            1.0                        1.0   
...              ...          ...            ...                        ...   
26703          53410          1.0            1.0                        0.0   
26704          53411          3.0            1.0                        0.0   
26705          53412          0.0            1.0                        0.0   
26706          53413          3.0            1.0                        0.0   
26707          53414          2.0            1.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                       1.0                   0.0                    1.0   
1                       0.0                   0.0                    0.0   
2                       0.0                   1.0                    1.0   
3                       0.0                   0.0                    0.0   
4                       1.0                   0.0                    1.0   
...                     ...                   ...                    ...   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   0.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    1.0   
26707                   0.0                   0.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
0                              1.0                      0.0   
1                              0.0                      0.0   
2                              1.0                      1.0   
3                              0.0                      0.0   
4                              1.0                      1.0   
...                            ...                      ...   
26703                          0.0                      0.0   
26704                          1.0                      1.0   
26705                          0.0                      0.0   
26706                          0.0                      1.0   
26707                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
0                        1.0  ...                  > $75,000     Not Married   
1                        0.0  ...              Below Poverty     Not Married   
2                        1.0  ...                  > $75,000         Married   
3                        0.0  ...  <= $75,000, Above Poverty         Married   
4                        1.0  ...  <= $75,000, Above Poverty     Not Married   
...                      ...  ...                        ...             ...   
26703                    1.0  ...                        NaN             NaN   
26704                    1.0  ...              Below Poverty         Married   
26705                    0.0  ...              Below Poverty     Not Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   
26707                    1.0  ...                        NaN     Not Married   

       rent_or_own   employment_status  hhs_geo_region  \
0             Rent            Employed        mlyzmhmf   
1             Rent            Employed        bhuqouqj   
2              Own            Employed        lrircsnp   
3              Own  Not in Labor Force        lrircsnp   
4              Own            Employed        lzgpxyit   
...            ...                 ...             ...   
26703          NaN   

In [6]:
train_feat.describe()

respondent_id   xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
count   26707.000000  26615.000000   26591.000000               26636.000000   
mean    13353.000000      1.618486       1.262532                   0.048844   
std      7709.791156      0.910311       0.618149                   0.215545   
min         0.000000      0.000000       0.000000                   0.000000   
25%      6676.500000      1.000000       1.000000                   0.000000   
50%     13353.000000      2.000000       1.000000                   0.000000   
75%     20029.500000      2.000000       2.000000                   0.000000   
max     26706.000000      3.000000       2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  ...  health_worker  health_insurance  \
count           26579.000000  ...   25903.000000       14433.00000   
mean                0.677264  ...       0.111918           0.87972   
std                 0.467531  ...       0.315271           0.32530   
min                 0.000000  ...       0.000000           0.00000   
25%                 0.000000  ...       0.000000           1.00000   
50%                 1.000000  ...       0.000000           1.00000   
75%                 1.000000  ...       0.000000           1.00000   
max                 1.000000  ...       1.000000           1.00000   

       opinion_xyz_vacc_effective  opinion_xyz_risk  \
count                26316.000000      26319.000000   
mean                     3.850623          2.342566   
std                      1.007436          1.285539   
min                      1.000000          1.000000   
25%                      3.000000          1.000000   
50%                      4.000000          2.000000   
75%                      5.000000          4.000000   
max                      5.000000          5.000000   

       opinion_xyz_sick_from_vacc  opinion_seas_vacc_effective  \
count                26312.000000                 26245.000000   
mean                     2.357670                     4.025986   
std                      1.362766                     1.086565   
min                      1.000000                     1.000000   
25%                      1.000000                     4.000000   
50%                      2.000000                     4.000000   
75%                      4.000000                     5.000000   
max                      5.000000                     5.000000   

       opinion_seas_risk  opinion_seas_sick_from_vacc  household_adults  \
count       26193.000000                 26170.000000      26458.000000   
mean            2.719162                     2.118112          0.886499   
std             1.385055                     1.332950          0.753422   
min        

In [7]:

#splitting the data


train_data = pd.merge(train_feat, train_label, on='respondent_id')

# Separate features and target variables
X = train_data.drop(columns=['respondent_id', 'xyz_vaccine', 'seasonal_vaccine'])
y_xyz = train_data['xyz_vaccine']
y_seasonal = train_data['seasonal_vaccine']
X_test = test_feat.drop(columns=['respondent_id'])



In [8]:
# Identifying different types of data we have 

#Categories 
categorical_cols = X.select_dtypes(include=['object']).columns

#Numerical Binary Columns - 0 or 1
binary_cols = [col for col in X.select_dtypes(include=['int64', 'float64']).columns if X[col].nunique() == 2]

#Normal Numerical Columns 
continuous_numerical_cols = [col for col in X.select_dtypes(include=['int64', 'float64']).columns if X[col].nunique() > 2]




In [9]:
# train_feat['behavioral_antiviral_meds']=train_feat['behavioral_antiviral_meds'].fillna(0)
# train_feat['behavioral_avoidance']=train_feat['behavioral_avoidance'].fillna(0)
# train_feat['behavioral_face_mask']=train_feat['behavioral_face_mask'].fillna(0)
# train_feat['behavioral_wash_hands']=train_feat['behavioral_wash_hands'].fillna(0)
# train_feat['behavioral_large_gatherings']=train_feat['behavioral_large_gatherings'].fillna(0)
# train_feat['behavioral_outside_home']=train_feat['behavioral_outside_home'].fillna(0)
# train_feat['behavioral_touch_face']=train_feat['behavioral_touch_face'].fillna(0)
# train_feat['doctor_recc_xyz']=train_feat['doctor_recc_xyz'].fillna(0)
# train_feat['doctor_recc_seasonal']=train_feat['doctor_recc_seasonal'].fillna(0)
# train_feat['chronic_med_condition']=train_feat['chronic_med_condition'].fillna(0)
# train_feat['child_under_6_months']=train_feat['child_under_6_months'].fillna(0)
# train_feat['health_worker']=train_feat['health_worker'].fillna(0)
# train_feat['health_insurance']=train_feat['health_insurance'].fillna(0)



In [10]:

train_feat.isna().sum()

respondent_id                      0
xyz_concern                       92
xyz_knowledge                    116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_xyz                 2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_xyz_vacc_effective       391
opinion_xyz_risk                 388
opinion_xyz_sick_from_vacc       395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4423
m

# Step 2 Defining Preprocessing pipelines for ease

In [11]:

continuous_numerical_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

binary_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value=0))
])

categorical_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

In [12]:
# Combining all the pipelines

preprocessor = ColumnTransformer(
    transformers=[
        ('num', continuous_numerical_pipeline, continuous_numerical_cols),
        ('bin', binary_pipeline, binary_cols),
        ('cat', categorical_pipeline, categorical_cols)
    ])

# Step 3 Defining our models 


### Used LogisticReg

In [13]:

model_xyz = LogisticRegression(max_iter=1000, random_state=42)
model_seasonal = LogisticRegression(max_iter=1000, random_state=42)


# Create separate pipelines for each target variable so that we can train them seperately

pipeline_xyz = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', model_xyz)
])

pipeline_seasonal = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', model_seasonal)
])

In [14]:
# Split data into training and validation sets for xyz
X_train_xyz, X_val_xyz, y_train_xyz, y_val_xyz = train_test_split(X, y_xyz, test_size=0.2, random_state=42)

# Train and predict for xyz
pipeline_xyz.fit(X_train_xyz, y_train_xyz)
y_val_pred_prob_xyz = pipeline_xyz.predict_proba(X_val_xyz)[:, 1]
roc_auc_xyz = roc_auc_score(y_val_xyz, y_val_pred_prob_xyz,average='macro')

In [15]:
# Split data into training and validation sets for seasonal
X_train_seasonal, X_val_seasonal, y_train_seasonal, y_val_seasonal = train_test_split(X, y_seasonal, test_size=0.3, random_state=42)

# Train and predict for seasonal
pipeline_seasonal.fit(X_train_seasonal, y_train_seasonal)
y_val_pred_prob_seasonal = pipeline_seasonal.predict_proba(X_val_seasonal)[:, 1]
roc_auc_seasonal = roc_auc_score(y_val_seasonal, y_val_pred_prob_seasonal,average='macro')

## Scores ROC AUC

In [16]:
print(f"ROC AUC for xyz_vaccine: {roc_auc_xyz:.4f}")
print(f"ROC AUC for seasonal_vaccine: {roc_auc_seasonal:.4f}")

print(f"Mean ROC AUC : {(roc_auc_seasonal+roc_auc_xyz)/2:.4f}")

ROC AUC for xyz_vaccine: 0.8600
ROC AUC for seasonal_vaccine: 0.8574
Mean ROC AUC : 0.8587


In [17]:
# Retrain the models on the entire training set
pipeline_xyz.fit(X, y_xyz)
pipeline_seasonal.fit(X, y_seasonal)

# Predict on the test set
y_test_pred_xyz = pipeline_xyz.predict_proba(X_test)[:, 1]
y_test_pred_seasonal = pipeline_seasonal.predict_proba(X_test)[:, 1]

# Step 4 Making the submission

In [18]:
# Prepare submission
submission = pd.DataFrame({
    'respondent_id': test_feat['respondent_id'],
    'xyz_vaccine': y_test_pred_xyz,
    'seasonal_vaccine': y_test_pred_seasonal
})

# Save to CSV
submission.to_csv('submission.csv', index=False)

In [19]:
y_test_pred_seasonal

array([0.33577134, 0.04619581, 0.54788043, ..., 0.22538468, 0.33640547,
       0.6137739 ])

In [20]:
y_test_pred_xyz

array([0.09225323, 0.01885194, 0.284527  , ..., 0.20481126, 0.03535098,
       0.78510075])

In [21]:
submission_df=pd.read_csv("submission.csv")

In [22]:
submission_df

respondent_id  xyz_vaccine  seasonal_vaccine
0              26707     0.092253          0.335771
1              26708     0.018852          0.046196
2              26709     0.284527          0.547880
3              26710     0.568397          0.893246
4              26711     0.301780          0.535014
...              ...          ...               ...
26703          53410     0.212825          0.484730
26704          53411     0.080587          0.332846
26705          53412     0.204811          0.225385
26706          53413     0.035351          0.336405
26707          53414     0.785101          0.613774

[26708 rows x 3 columns]

# Result 

## Mean ROC AUC of seasonal and xyz vaccine predictions = 0.8587
